internal/spoke_k8s_client/factory.go

```diff
@@ -1,90 +1,255 @@
+		return nil, err
+	}
+	result, ok := secret.Data["kubeconfig"]
+	if !ok || len(result) == 0 {
+		err := errors.Errorf(
+			"secret '%s/%s' doesn't contain the 'kubeconfig' key",
+			secret.Namespace, secret.Name,
+		)
+		return nil, err
+	}
+	return result, nil
+}
+
+func (f *spokeK8sClientFactory) clientFromRestConfig(restConfig *rest.Config) (ctrlclient.Client, error) {
 	schemes := GetKubeClientSchemes()
-	targetClient, err := client.New(restConfig, client.Options{Scheme: schemes})
+	return ctrlclient.New(restConfig, ctrlclient.Options{Scheme: schemes})
+}
+
+func (f *spokeK8sClientFactory) clientSetFromRestConfig(restConfig *rest.Config) (*kubernetes.Clientset, error) {
+	return kubernetes.NewForConfig(restConfig)
+}
+
+// isFromHostedCluster determines if this cluster corresponding to the given kubeconfig secret is a hosted cluster.
+// Returns a boolean flag and, when that flag is true, the namespace and name of the cluster deployment of the hosted
+// cluster.
+func (f *spokeK8sClientFactory) isFromHostedCluster(ctx context.Context, kubeconfigSecret *corev1.Secret) (bool,
+	types.NamespacedName, error) {
+	logger := f.logger.WithFields(logrus.Fields{
+		"namespace": kubeconfigSecret.Namespace,
+		"name":      kubeconfigSecret.Name,
+	})
+
+	// Try to find the cluster deployment. If we can't, for whatever the reason, explain it in the log and assume
+	// it isn't a hosted cluster.
+	clusterDeployment, err := f.findClusterDeploymentForKubeconfigSecret(ctx, kubeconfigSecret)
 	if err != nil {
-		cf.log.WithError(err).Warnf("failed to get spoke kube client")
-		return nil, nil, err
+		logger.WithError(err).Warn(
+			"Failed to find cluster deployment corresponding to kubeconfig secret, will assume it isn't " +
+				"a hosted cluster",
+		)
+		return false, types.NamespacedName{}, err
+	}
+	if clusterDeployment == nil {
+		logger.Warn(
+			"Cluster deployment corresponding to kubeconfig secret doesn't exist, will assume it isn't " +
+				"a hosted cluster",
+		)
+		return false, types.NamespacedName{}, err
+	}
+
+	// We assume this is a hosted cluster if it has the 'agentClusterRef' label, no matter the value.
+	_, result := clusterDeployment.Labels["agentClusterRef"]
+	var clusterDeploymentKey types.NamespacedName
+	if result {
+		clusterDeploymentKey.Namespace = clusterDeployment.Namespace
+		clusterDeploymentKey.Name = clusterDeployment.Name
 	}
+	return result, clusterDeploymentKey, nil
+}
 
-	spokeClient := &spokeK8sClient{
-		Client:      targetClient,
-		csrClient:   clientset.CertificatesV1().CertificateSigningRequests(),
-		sarClient:   clientset.AuthorizationV1().SelfSubjectAccessReviews(),
-		nodesClient: clientset.CoreV1().Nodes(),
-		log:         cf.log,
+// findClusterDeploymentForKubeconfigSecret finds the cluster deployment that corresponds to the given kubeconfig
+// secret. It returns nil if there is no such cluster deployment.
+func (f *spokeK8sClientFactory) findClusterDeploymentForKubeconfigSecret(ctx context.Context,
```

Comment:
Are we ever in a situation where the caller if this factory doesn't already have a reference to the cluster deployment?

Since (based on the naming) we're talking about "spoke" clusters it seems likely that this could be simplified by either the caller supplying the cluster deployment or by this logic living outside this factory (then we would have an option like "useHubServiceNetwork" or something when creating the client).

Reply:
Yes, here we don't know what is the cluster deployment: https://github.com/openshift/assisted-service/blob/9a1b9eceb81c43fb0655022d393fdf9911dd537c/internal/controller/controllers/hypershiftagentserviceconfig_controller.go#L342 .

Reply:
Okay, thanks.

Side note though ... can we delete the HASC CRD and controller yet?
@gamli75 that effort isn't happening now, right?

Reply:
I'm not familiar with that effort. maybe @CrystalChun

Reply:
I'm not familiar with it either, maybe @danielerez?